platform/preferences/preferences-impl/src/main/java/org/codice/ddf/preferences/impl/PreferencesImpl.java

```diff
@@ -0,0 +1,227 @@
+
+  /** These are the attributes that will be ignored from being added to the preferences metacard. */
+  private static final Set<String> IGNORED_ATTRIBUTES =
+      new HashSet<>(Arrays.asList(PreferencesMetacardType.USER_ATTRIBUTE, Metacard.ID));
+
+  private final List<PostRetrievalPlugin> postRetrievalPlugins;
+
+  private final List<DefaultPreferencesSupplier> defaultPreferencesSuppliers;
+
+  private final List<InjectableAttribute> injectableAttributes;
+
+  public PreferencesImpl(
+      CatalogFramework catalogFramework,
+      FilterBuilder filterBuilder,
+      List<PostRetrievalPlugin> postRetrievalPlugins,
+      List<DefaultPreferencesSupplier> defaultPreferencesSuppliers,
+      List<InjectableAttribute> injectableAttributes) {
+    this.catalogFramework = catalogFramework;
+    this.filterBuilder = filterBuilder;
+    this.postRetrievalPlugins = postRetrievalPlugins;
+    this.defaultPreferencesSuppliers = defaultPreferencesSuppliers;
+    this.injectableAttributes = injectableAttributes;
+  }
+
+  @Override
+  public void add(Map<String, Object> preferences, String userId) throws PreferencesException {
+
+    Filter filter = createFilter(userId);
+
+    QueryResponse queryResponse = query(filter);
+
+    if (queryResponse.getResults().size() > 1) {
+      throw new PreferencesException("internal error: more than one preference metacard found");
+    } else if (queryResponse.getResults().size() == 1) {
+      updateExistingMetacard(preferences, queryResponse);
+    } else {
+      createNewMetacard(preferences, userId);
+    }
+  }
+
+  private QueryResponse query(Filter filter) throws PreferencesException {
+    QueryResponse queryResponse;
+    try {
+      queryResponse = catalogFramework.query(new QueryRequestImpl(new QueryImpl(filter)));
+    } catch (UnsupportedQueryException | SourceUnavailableException | FederationException e) {
+      throw new PreferencesException(e);
+    }
+    return queryResponse;
+  }
+
+  private Filter createFilter(String userId) {
+    return filterBuilder.allOf(
+        filterBuilder.attribute(PreferencesMetacardType.USER_ATTRIBUTE).is().equalTo().text(userId),
+        filterBuilder.attribute(Metacard.TAGS).is().equalTo().text(PreferencesMetacardType.TAG));
+  }
+
+  private void createNewMetacard(Map<String, Object> preferences, String userId)
+      throws PreferencesException {
+    Metacard metacard = new MetacardImpl(METACARD_TYPE);
+    metacard.setAttribute(new AttributeImpl(PreferencesMetacardType.USER_ATTRIBUTE, userId));
+    metacard.setAttribute(new AttributeImpl(Metacard.TAGS, PreferencesMetacardType.TAG));
+    // TODO I could loop through the incoming preferences or loop through the known attributes, not
+    // sure if one way is better than the other
+    preferences.forEach(
+        (key, value) -> {
+          if (!IGNORED_ATTRIBUTES.contains(key)) {
+            String json = GSON.toJson(value);
+            metacard.setAttribute(new AttributeImpl(key, json));
```

Comment:
Why not just have a single attribute that holds the whole map as a json blob? Seems like it would make it simpler to work with. No need to worry about ignored attributes or attribute injectors.

Reply:
No additional attribute injector configuration needed either for new fields in the map

Reply:
the reason for making them separate is to make it easier to examine the configuration directly in solr and to allow downstream projects (or new modules) to add their own attributes.